### PR TITLE
Add dataset implementations for codrep and bug2fix

### PR DESCRIPTION
**Pull request by gchaperon**
This commit adds 2 things. First, implementations for the codrep and
bug2fix datasets. I tried using a similar interface to the datasets
found in torchvision and pytorch in general.
Second, I refactored the code for generating the buggy context, so that
it works better with JPype in python. JPype doesn't manage System.exit()
very well, and also it's super easy to send String(s) to the JVM using
JPype, so I changed the signatures of the relevant methods in the java
code.

### CODE REVIEW

**Comment by gchaperon** [comment]
@chenzimin 
I would very much appreciate your comment on the preprocessing steps, specially the filtering of single line patches in bug2fix.
Is what I have implemented relatively similar to what you had done for sequencer?

**Comment by chenzimin** [comment]
Hi,

It all looks correct! But one thing that I remember that I needed to do on the CodRep dataset is to remove comment changes, ie. removing changes that adds, removes, or replaces comments. For the comment dataset, I checked if the inserted or removed line starts with `//`. This will handle inline comments. For multi-line comments, I checked with gumtree that the diffs actually contains a AST diff.

The code that I check for AST diff is:
```java
    public static boolean contains_AST_diffs(File pre, File post) {
    		Diff diff;
    		try {
    				diff = new AstComparator().compare(pre, post);
			} catch (Exception e) {
				return false;
			}

    		List<Operation> operations = diff.getAllOperations();
    		if(operations.size() == 0)
    		{
    			return false;
    		}

    		CtElement el = diff.commonAncestor();
    		if(el instanceof CtAnnotation) {
        		return false;
        }

    		return true;
    }
```

**Comment by gchaperon** [comment]
hi!

I thought that is already handled by the buggy context abstraction or no?  I am thinking of [this](https://github.com/KTH/chai/blob/4eb2fd90560e93a2121b643cfe96b5babb5f1939/src/Buggy_Context_Abstraction/abstraction/src/main/java/se/kth/methodWithClassContextAbstraction.java#L84).
I thought of allowing one line fixes to comment in the first filtering step (for bug2fix), and when producing the merged dataset use the context abstraction to filter unsupported one line fixes.

let me know what you think

**Comment by gchaperon** [comment]
Hi Zimin,

as discussed last week I filtered single line fixes that include comments, so lines that contain `//`.
If I understood correctly that was the only case left after filtering out multiline fixes (done in bug2fix dataset implementation) and buggy lines that contain comments (done by abstraction java package).

I admit I have been reluctant to use gumtree to check diffs, mostly because the bug2fix zip file contains ~1.5M buggy files, and the preprocessing of those using java would have been really slow. Right now the filtering of bug2fix files is only text processing, it happens at around ~550 files/s,  and it takes around 45 min.

The relevant filtering steps are:
1. Filter bug2fix files with only single line diffs. The criterion is defined [here](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/datasets/bug2fix.py#L38)
2. Keep all CodRep split unchanged
3. For test set of sequencer dataset un split 4 of codrep, for train set of sequencer concat splits 1, 2, 3 and 5 of codrep and bug2fix data. This can be found[ here](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/datasets/sequencer.py#L169)
4. Apply filtering to __all__ examples (both datasets). First remove blacklisted files (turns out `spoon` has a bug where it gets stuck in a infinite loop, this happends for certain files), then remove examples that contain comments in the fixed line. Found [here ](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/datasets/sequencer.py#L186)
5. Perform buggy context abstraction. This step also filters some examples, mainly the ones where the buggy line contains either a comment or an annotation. After this step it is guaranteed that the abstracted source contains the marking comment, so no need to further check that after. Logic can be found [here](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/abstraction/src/main/java/se/kth/abstraction/ContextAbstraction.java#L62)

All examples from codrep and bug2fix sum up to ~250k, abstraction happens at ~100files/s, so around 45 min if both datasets were cached before. After abstraction there are ~150k examples left.

Also, if it's ok with you, I will be hosting the data after all preprocessing at  https://users.dcc.uchile.cl/~gchapero/datasets/sequencer-data.tar.gz, so that it is easier to work with when mixing local development with running in different servers.
With this, instantiating the data is as simple as 
```python
dataset = SequenceRDataset.from_tar("data_path", split="train")
# data_path is for saving cached files and such, not the location of the tar.gz file
```

Let me know what you think!
Best,
Gabriel

**Comment by gchaperon** [comment]
hi,
[here](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/datasets/bug2fix.py#L56) I obtain t he buggy/fixed pairs. So [this](https://github.com/KTH/chai/blob/b779e60fa6bb2fbb6a1dea602a5e3a52600aa9e5/src/embedding-work/codebert/datasets/bug2fix.py#L95) diff is only for two files, not the whole commit. Now, I had to check `len(patch) >= 1` so that empty diff are ignored, and also I though it would be impossible to get a multifile patch from the  diff of a single file, but since I am not sure I set it to `len(patch) == 1`. Now that I think about it, iterating over `patch` handles both cases better (when it is empty and when there are multiple files). I will change this.

**Comment by gchaperon** [comment]
I just added your recommendation, I rerun the preprocessing and nothing changed. I guess it is a matter of preference how to write it.

**Comment by gchaperon** [comment]
I also uploaded the data to zenodo and added the new link in the source code.

**Comment by chenzimin** [comment]
Thanks for the PR!
